File: UVA_10484_DivisibilityOfFactors.java
import java.io.*;
import java.util.Scanner;

public class UVA_10484_DivisibilityOfFactors 
{
  public static void main(String[] args) 
	{
		int i, j;
		int[] primeFactor = new int[101];
		int[] primes = new int[101];
		int counterp=0;
		
		for(i=0; i<=100; i++)
			primeFactor[i] = 0;
		
		for(i=2; i*i<=100; i++)
		{
			if(primeFactor[i] == 0)
			{
				primes[counterp] = i;
				counterp++;
				primeFactor[i] = i;
				for(j=i*i; j<=100; j+=i)
					if(primeFactor[j] == 0)
						primeFactor[j] = i;
			}
		}
		
		for(j=i; j<=100; j++)
			if(primeFactor[j] == 0)
			{
				primeFactor[j] = j;
				primes[counterp] = j;
				counterp++;
			}
		
		Scanner in = new Scanner(System.in);
		int N = in.nextInt();
		long D = in.nextLong();
		while(N!=0 || D!=0)
		{
			int countN[] = new int[101];
			int countD[] = new int[101];
			int tempN = N;
			long tempD;
			long result = 1;
			int flag = 1;
			if(D>0)
				tempD = D;
			else
				tempD = -D;
			for(int k=N; k>1; k--)
			{
				tempN = k;
				while(tempN != 1)
				{
					countN[primeFactor[tempN]] ++;
					tempN = tempN / primeFactor[tempN];
				}
			}
			int counter=0;
			while(tempD!=1)
			{
				if(tempD % primes[counter] == 0)
				{
					countD[primes[counter]] ++;
					tempD = tempD / primes[counter];
				}
				else if(counter+1 < counterp)
					counter++;
				else 
				{
					flag = 0;
					break;
				}
			}
			for(i=0; i<100; i++)
			{
				if(countN[i] > countD[i])
					result *=  (countN[i] - countD[i] + 1);
				else if(countN[i] < countD[i])
				{
					flag = 0;
					break;
				}
			}
			if(flag==1)
				System.out.println(result);
			else if(flag==0)
				System.out.println(0);
			N = in.nextInt();
			D = in.nextInt();
		}
	}
}

File: UVA_10484_DivisibilityOfFactors.cpp
#include <iostream>
#include <vector>
#include <map>
#include <math.h>
using namespace std;

vector<int> primes;
bool isPrime[1000];
void sieve(int maxN)
{
    int i, j;
    for(int i=0; i<=maxN; i++)
    {
        isPrime[i] = true;
    }
    isPrime[0] = false;
    isPrime[1] = false;
    for(i=2; i*i<=maxN; i++)
    {
        if(isPrime[i] == true)
        {
            primes.push_back(i);
            //cout << i << endl;
            for(j=i*i; j<=maxN; j+=i)
            {
                isPrime[j] = false;
            }
        }
    }
    for(i=i; i<=maxN; i++)
    {
        if(isPrime[i] == true)
        {
            primes.push_back(i);
            //cout << i << endl;
        }
    }
}

int main()
{
    int N, D, tempD, tempN, counter;
    vector<int> primeFactors_D;
    map<int, int> primeFactorsPower_D;
    vector<int> primeFactors_N;
    map<int, int> primeFactorsPower_N;

    sieve(1000);

    cin >> N >> D;
    while(N != 0 || D != 0)
    {
        tempD = D;
        tempN = N;

        // Factorizing D
        primeFactors_D.clear();
        primeFactorsPower_D.clear();
        for(int i=2; i*i<=D; i++)
        {
            if(tempD%i == 0)
            {
                primeFactors_D.push_back(i);
                primeFactorsPower_D[i] = 0;
                while(tempD % i == 0)
                {
                    tempD = tempD / i;
                    primeFactorsPower_D[i] ++;
                }
                //cout << i << " " << primeFactorsPower_D[i] << endl;
            }
        }
        if(tempD != 1)
        {
            primeFactors_D.push_back(tempD);
            primeFactorsPower_D[tempD] = 1;
            //cout << tempD << " " << primeFactorsPower_D[tempD] << endl;
        }

        // Factorizing N!
        primeFactors_N.clear();
        primeFactorsPower_N.clear();
        for(int pi=0; pi<primes.size(); pi++)
        {
            int nowPower = 1;
            int toAdd = tempN / pow(primes[pi], nowPower);
            if(toAdd == 0)
            {
                break;
            }
            if(toAdd != 0)
            {
                primeFactors_N.push_back(primes[pi]);
                primeFactorsPower_N[primes[pi]] = 0;
                while(toAdd != 0)
                {
                    //cout << toAdd << endl;
                    primeFactorsPower_N[primes[pi]] += toAdd;
                    nowPower ++;
                    toAdd = tempN / pow(primes[pi], nowPower);
                }
                //cout << primes[pi] << " " << primeFactorsPower_N[primes[pi]] << endl;
            }
        }

        long long result = 1;
        // Checking that all the D Prime Factors are found in N! Prime Factors
        for(int i=0; i<primeFactors_D.size(); i++)
        {
            int nowPrimeNumber = primeFactors_D[i];
            if(primeFactorsPower_N[nowPrimeNumber] < primeFactorsPower_D[nowPrimeNumber])
            {
                result = 0;
            }
        }

        // Calculating the Result
        for(int i=0; i<primeFactors_N.size(); i++)
        {
            if(primeFactorsPower_D.count(primeFactors_N[i]) != 0)
            {
                result *= (primeFactorsPower_N[primeFactors_N[i]] + 1) - primeFactorsPower_D[primeFactors_N[i]];
            }
            else
            {
                result *= (primeFactorsPower_N[primeFactors_N[i]] + 1) - 0;
            }
        }

        cout << result << endl;

        cin >> N >> D;
    }

    return 0;
}
